scalding-core/src/main/scala/com/twitter/scalding/typed/HashJoinable.scala

```diff
@@ -50,11 +51,53 @@ trait HashJoinable[K, +V] extends CoGroupable[K, V] with KeyedPipe[K] {
       val newPipe = new HashJoin(
         RichPipe.assignName(mapside.toPipe(('key, 'value))(fd, mode, tup2Setter)),
         Field.singleOrdered("key")(keyOrdering),
-        mapped.toPipe(('key1, 'value1))(fd, mode, tup2Setter),
+        getMappedPipe(fd, mode),
         Field.singleOrdered("key1")(keyOrdering),
         WrappedJoiner(new HashJoiner(joinFunction, joiner)))
 
       //Construct the new TypedPipe
       TypedPipe.from[(K, R)](newPipe.project('key, 'value), ('key, 'value))(fd, mode, tuple2Converter)
     })
+
+  /**
+   * Returns a Pipe for the mapped (rhs) pipe with checkpointing applied if needed.
+   * Currently we skip checkpointing if:
+   * 1) If the pipe is a ForceToDiskTypedPipe (and thus the Cascading pipe is a Checkpoint)
+   * 2) If the Cascading pipe is a GroupBy / CoGroup
+   * 3) If while walking the pipe's parents we hit an Every before an Each
+   */
+  private def getMappedPipe(fd: FlowDef, mode: Mode): Pipe = {
+    val mappedPipe = mapped.toPipe(('key1, 'value1))(fd, mode, tup2Setter)
+
+    mappedPipe match {
+      case _: Checkpoint => mappedPipe
+      case _: GroupBy => mappedPipe
+      case _: CoGroup => mappedPipe
+      case _ =>
+        val checkParents = checkEveryBeforeEach(mappedPipe)
+        checkParents match {
+          case Some(true) => mappedPipe
+          case _ => mappedPipe.forceToDisk
+        }
+    }
+  }
+
+  /**
+   * Check if we encounter an Every before an Each while traversing parents of the pipe
+   */
+  private def checkEveryBeforeEach(pipe: Pipe): Option[Boolean] = {
```

Comment:
Yeah, I've updated this based on your suggestion below with the .collect.toList ...